### DISCUSSION TITLE
Validation for SG Policies, should return SG-Rule violated along with current resource response

### DISCUSSION THREAD

**Discussion by mihirv0ra** [cloud-custodian/cloud-custodian#7312] [General]
The current SG validations work as expected for defined polices, however the response of the Custodian validation for a particular policy comprises of the following,
```
[
  {
    "VpcId": "vpc-xxxxxxxx",
    "Description": "Security Group for xxxxxxx",
    "Tags": [
      {
        "Value": "TEST-VALUE-FLAG-FAIL",
        "Key": "TEST-TAG"
      },
      {
        "Value": "TEST-XXXX",
        "Key": "Name"
      }
    ],
    "IpPermissions": [
      {
        "PrefixListIds": [],
        "FromPort": 0,
        "IpRanges": [],
        "ToPort": 0,
        "IpProtocol": -1,
        "UserIdGroupPairs": [
          {
            "UserId": "XXXXXXXX",
            "GroupId": "XXXXXXXX"
          }
        ]
      },
      {
        "ToPort": 8443,
        "FromPort": 8443,
        "PrefixListIds": [],
        "IpRanges": [
          {
            "CidrIp": "0.0.0.0/0"
          },
          {
            "CidrIp": "10.0.0.0/0"
          }
        ],
        "IpProtocol": "tcp"
      }
    ],
    "GroupName": "TEST-XXXX",
    "IpPermissionsEgress": [
      {
        "PrefixListIds": [],
        "FromPort": 0,
        "IpRanges": [],
        "ToPort": 0,
        "IpProtocol": -1,
        "UserIdGroupPairs": [
          {
            "UserId": "XXXXXXXX",
            "GroupId": "XXXXXXXX"
          }
        ]
      }
    ],
    "MatchedFilters": [
      "tag:TEST-TAG"
    ],
    "OwnerId": "<computed>",
    "MatchedIpPermissions": [
      {
        "ToPort": 8443,
        "FromPort": 8443,
        "PrefixListIds": [],
        "IpRanges": [
          {
            "CidrIp": "0.0.0.0/0"
          }
        ],
        "IpProtocol": "tcp"
      },
      {
        "ToPort": 8443,
        "FromPort": 8443,
        "PrefixListIds": [],
        "IpRanges": [
          {
            "CidrIp": "10.0.0.0/0"
          }
        ],
        "IpProtocol": "tcp"
      }
    ],
    "GroupId": "XXXXXXXX"
  }
]
```

It lists MatchedIPPermissions with one element for each CIDR, and it is hard to get what rule failed from that, with the partial information, also the current output creates some confusion incase the same CIDR blocks are defined on multiple rules

A suggestion would be to list the mapping of the sg-rules, corresponding to the Policy and send the rule in the response along with the current fields, so adding another element like MatchedRules and then having the same populated with the rule definition for quick lookup,

```
"MatchedRules" :[
   {
        "ToPort": 8443,
        "FromPort": 8443,
        "PrefixListIds": [],
        "IpRanges": [
          {
            "CidrIp": "0.0.0.0/0"
          },
          {
            "CidrIp": "10.0.0.0/0"
          }
        ],
        "IpProtocol": "tcp"
      }
] 
```

**Comment by kapilt** [comment]
we expanded the rules because it matches the unit of modification for the common case as the describe group output automatically collapses multiple perms by port & protocol, the modification for addition/deletion is more granular then that though, the expansion allows fine grained pinpointing of the exact permission that causes a match, it also corresponds to a set of permissions that can be used with modify apis. i'm a little unclear why you would want the collapsed version, say you matched a cidr like 0.0.0.0/0  would you also want it to show a permission that contained 10.0.0.0/8 ... if you visit the aws console.. you'll see these are displayed separately.. could you elaborate on why you want this behavior, is it some sort of ui thing, also a sample policy would be helpful.
